### ISSUE TITLE
Refactoring class ResnetGenerator(nn.Module) in networks.py

### ISSUE THREAD

**Issue by omrysendik** [junyanz/pytorch-CycleGAN-and-pix2pix#308]
Hi,

Please forgive my perhaps basic question. I am a newbie in PyTorch...
I would like to refactor the class ResnetGenerator(nn.Module) in networks.py so that instead of continuously applying "model+=" in order to build the model, I define an Encoder and Decoder.
That's because I would like to save the output of the Encoder.
The problem is that applying my changes yields a model with an absurdly small amount of parameters (the number is printed through "def print_networks(self, verbose):" within base_model.py).
Note that I would like to apply a .view operation on the Enc output tensor, meaning that I can't use register_hook.

Help is greatly appreciated.

The change I am making looks as follows:

class ResnetGenerator(nn.Module):
    def __init__(self, input_nc, output_nc, ngf=64, norm_layer=nn.BatchNorm2d, use_dropout=False, n_blocks=6, padding_type='reflect'):
        assert(n_blocks >= 0)
        super(ResnetGenerator, self).__init__()
        self.input_nc = input_nc
        self.output_nc = output_nc
        self.ngf = ngf
        if type(norm_layer) == functools.partial:
            use_bias = norm_layer.func == nn.InstanceNorm2d
        else:
            use_bias = norm_layer == nn.InstanceNorm2d

        Enc = [nn.ReflectionPad2d(3),
                 nn.Conv2d(input_nc, ngf, kernel_size=7, padding=0,
                           bias=use_bias),
                 norm_layer(ngf),
                 nn.ReLU(True)]

        n_downsampling = 2
        for i in range(n_downsampling):
            mult = 2**i
            Enc += [nn.Conv2d(ngf * mult, ngf * mult * 2, kernel_size=3,
                                stride=2, padding=1, bias=use_bias),
                      norm_layer(ngf * mult * 2),
                      nn.ReLU(True)]

        mult = 2**n_downsampling
        for i in range(n_blocks):
            Enc += [ResnetBlock(ngf * mult, padding_type=padding_type, norm_layer=norm_layer, use_dropout=use_dropout, use_bias=use_bias)]

        for i in range(n_downsampling):
            mult = 2**(n_downsampling - i)
            if i==0:
                 Dec = [nn.ConvTranspose2d(ngf * mult, int(ngf * mult / 2),
                                         kernel_size=3, stride=2,
                                         padding=1, output_padding=1,
                                         bias=use_bias),
                      norm_layer(int(ngf * mult / 2)),
                      nn.ReLU(True)]
            else:
                 Dec += [nn.ConvTranspose2d(ngf * mult, int(ngf * mult / 2),
                                         kernel_size=3, stride=2,
                                         padding=1, output_padding=1,
                                         bias=use_bias),
                      norm_layer(int(ngf * mult / 2)),
                      nn.ReLU(True)]
        Dec += [nn.ReflectionPad2d(3)]
        Dec += [nn.Conv2d(ngf, output_nc, kernel_size=7, padding=0)]
        Dec += [nn.Tanh()]

    def forward(self, input):
        Feat = self.Enc(input)
        return self.Dec(Feat)

**Comment by omrysendik** [comment]
Solved. Thanks.
